hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -993,6 +993,244 @@ class Emit[C](
+          val K = new Value[Long] {
+            def get: Code[Long] = (M < N).mux(M, N)
+          }
+          val LDA = new Value[Long] {
+            override def get: Code[Long] = (M > 1L).mux(M, 1L) // Possible stride tricks could change this in the future.
+          }
+
+          def LWORK = Region.loadDouble(LWORKAddress).toI
+
+          val dataAddress = pndValue.pt.data.load(pndValue.tcode[Long])
+
+          val tauPType = PCanonicalArray(PFloat64Required, true)
+          val tauAddress = mb.genFieldThisRef[Long]()
+          val workAddress = mb.genFieldThisRef[Long]()
+          val aAddressDGEQRF = mb.genFieldThisRef[Long]() // Should be column major
+          val rDataAddress = mb.genFieldThisRef[Long]()
+          val aNumElements = mb.genFieldThisRef[Long]()
+
+          val infoDGEQRFResult = mb.newLocal[Int]()
+          val infoDGEQRFErrorTest = (extraErrorMsg: String) => (infoDGEQRFResult cne 0)
+            .orEmpty(Code._fatal[Unit](const(s"LAPACK error DGEQRF. $extraErrorMsg Error code = ").concat(infoDGEQRFResult.toS)))
+
+          val computeHAndTau = Code(FastIndexedSeq(
+            aNumElements := pndValue.pt.numElements(shapeArray, mb),
+
+            // Make some space for A, which will be overriden during DGEQRF
+            aAddressDGEQRF := pndValue.pt.data.pType.allocate(region.code, aNumElements.toI),
+            pndValue.pt.data.pType.stagedInitialize(aAddressDGEQRF, aNumElements.toI),
+            Region.copyFrom(pndValue.pt.data.pType.firstElementOffset(dataAddress, (M * N).toI),
+              pndValue.pt.data.pType.firstElementOffset(aAddressDGEQRF, aNumElements.toI), (M * N) * 8L),
+
+            tauAddress := tauPType.allocate(region.code, K.toI),
+            tauPType.stagedInitialize(tauAddress, K.toI),
+
+            LWORKAddress := region.code.allocate(8L, 8L),
+
+            infoDGEQRFResult := Code.invokeScalaObject7[Int, Int, Long, Int, Long, Long, Int, Int](LAPACK.getClass, "dgeqrf",
+              M.toI,
+              N.toI,
+              pndValue.pt.data.pType.firstElementOffset(aAddressDGEQRF, aNumElements.toI),
+              LDA.toI,
+              tauPType.firstElementOffset(tauAddress, K.toI),
+              LWORKAddress,
+              -1
+            ),
+            infoDGEQRFErrorTest("Failed size query."),
+
+            workAddress := Code.invokeStatic1[Memory, Long, Long]("malloc", LWORK.toL * 8L),
+
+            infoDGEQRFResult := Code.invokeScalaObject7[Int, Int, Long, Int, Long, Long, Int, Int](LAPACK.getClass, "dgeqrf",
+              M.toI,
+              N.toI,
+              pndValue.pt.data.pType.firstElementOffset(aAddressDGEQRF, aNumElements.toI),
+              LDA.toI,
+              tauPType.firstElementOffset(tauAddress, K.toI),
+              workAddress,
+              LWORK
+            ),
+
+            Code.invokeStatic1[Memory, Long, Unit]("free", workAddress.load()),
+            infoDGEQRFErrorTest("Failed to compute H and Tau.")
+          ))
```

Comment:
You use `computeHAndTau` in both branches, this indicates that the setup could just exist on the CodeBuilder.